src/Haskell/Ide/Engine/Plugin/HsImport.hs

```diff
@@ -224,61 +323,174 @@ codeActionProvider plId docId _ context = do
-  mkImportAction importList diag modName = do
+    :: ModuleName -> ImportDiagnostic -> Maybe SymbolKind -> IdeM (Maybe J.CodeAction)
+  mkImportAction modName importDiagnostic symbolType = do
     cmd <- mkLspCommand plId "import" title (Just cmdParams)
     return (Just (codeAction cmd))
-   where
+    where
     codeAction cmd = J.CodeAction title
                                   (Just J.CodeActionQuickFix)
-                                  (Just (J.List [diag]))
+                                  (Just (J.List [diagnostic importDiagnostic]))
                                   Nothing
                                   (Just cmd)
-    title =
-      "Import module "
-        <> modName
-        <> maybe "" (\name -> " (" <> name <> ")") importList
-    cmdParams = [toJSON (ImportParams (docId ^. J.uri) importList modName)]
+    title = "Import module "
+      <> modName
+      <> case termType importDiagnostic of
+        Hiding _ -> "hiding"
+        -- ^ Note, that it must never happen
+        -- in combination with `symbolType == Nothing`
+        Import _ -> ""
+      <> case symbolType of
+        Just s  -> case s of
+          Only sym  -> " (" <> sym <> ")"
+          AllOf dt -> " (" <> dt <> " (..))"
+          OneOf dt sym -> " (" <> dt <> " (" <> sym <> "))"
+        Nothing -> ""
+
+    importStyleParam :: ImportStyle
+    importStyleParam = case symbolType of
+        Nothing -> Simple
+        Just k -> case termType importDiagnostic of
+          Hiding _ -> Complex (Hiding k)
+          Import _ -> Complex (Import k)
+
+    cmdParams = [toJSON (ImportParams (docId ^. J.uri) importStyleParam modName)]
 
 
   -- | For a Diagnostic, get an associated function name.
   -- If Ghc-Mod can not find any candidates, Nothing is returned.
-  getImportables :: J.Diagnostic -> Maybe (J.Diagnostic, T.Text)
+  getImportables :: J.Diagnostic -> Maybe ImportDiagnostic
   getImportables diag@(J.Diagnostic _ _ _ (Just "ghcmod") msg _) =
-    (diag, ) <$> extractImportableTerm msg
+    uncurry (ImportDiagnostic diag) <$> extractImportableTerm msg
   getImportables _ = Nothing
 
 -- | Extract from an error message an appropriate term to search for.
 -- This looks at the error message and tries to extract the expected
 -- signature of an unknown function.
 -- If this is not possible, Nothing is returned.
-extractImportableTerm :: T.Text -> Maybe T.Text
-extractImportableTerm dirtyMsg = T.strip <$> asum
-  [ T.stripPrefix "Variable not in scope: " msg
-  , T.init <$> T.stripPrefix "Not in scope: type constructor or class ‘" msg
-  , T.stripPrefix "Data constructor not in scope: " msg
-  ]
- where
-  msg =
-    head
-        -- Get rid of the rename suggestion parts
+extractImportableTerm :: T.Text -> Maybe (T.Text, SymbolImport SymbolType)
+extractImportableTerm dirtyMsg = do
+  (n, s) <- extractedTerm
+  let n' = T.strip n
+  return (n', s)
+  where
+    importMsg = head
```

Comment:
I’m always uncomfortable when I see partial functions. Can you use Safe.headMay?

Reply:
I thought the same, but it seems to never have exploded onto someone. I will change it.

Reply:
@fendor I understand we are waiting for you to change this before we can merge?

Reply:
Done now.